main/mini-geany.c

```diff
@@ -333,13 +342,12 @@ extern int main (int argc, char **argv)
 		{
 			printf("\nParsing %s:\n", argv[i]);
 			/* createTags() is called repeatadly during Geany execution */
-			createTags(NULL, 0, argv[i], getNamedLanguage("C", 0));
+			tagArray = createTags(NULL, 0, argv[i], getNamedLanguage("C", 0));
```

Comment:
Getting rid of using the static global variable is definitely nice. However, while this concrete approach works for this example, it isn't the best interface for Geany. When we call createTags(), we already have the tag array allocated inside TMSourceFile and we need to get the existing array filled and not newly created inside preWriteEntry() and then returned by createTags().

However, I think we already have all we need. We can just call

```
customWriter->private = our_tag_array; //or whatever we want
createTags(NULL, 0, argv[i], getNamedLanguage("C", 0));
```

and pass our custom data in and out this way without any return value of createTags(). The 'result' can be removed from postWriteEntry() too.

Reply:
I see.
However, providing the way to return results from a custom writer is a good idea.
Though, this time is not generalizing, I would like to introduce this chanage of *ctags side* separately.
I will make another pull request for this change.

I will not change mini-geany side.

Reply:
> However, providing the way to return results from a custom writer is a good idea.

I'm not sure if it will ever be useful for anything since `private` can hold arbitrary user data (e.g. a struct, in Geany we'll pass TMSourceFile using it, see https://github.com/geany/geany/blob/master/src/tagmanager/tm_source_file.h#L32). Do you have any example when this might be needed?

Reply:
The `private` field is designed as the place store data having rather short life type.
It is initialized with preWriteEntry method called from parseFileWithMio.
and set to NULL in postWriteEntry method called from parseFileWithMio.
This is designed for etags writer.
See postWriteEntry method of etags writer. It tells you much more than my English.

```
extern void writerSetup (MIO *mio)
{
	if (writer->preWriteEntry)
		writer->private = writer->preWriteEntry (writer, mio);
	else
		writer->private = NULL;
}

extern bool writerTeardown (MIO *mio, const char *filename)
{
	if (writer->postWriteEntry)
	{
		bool r;
		r = writer->postWriteEntry (writer, mio, filename);
		writer->private = NULL;
		return r;
	}
	return false;
}
```
Of course, you can bypass the life-cycle management done by ctags.
However, it is a kind of abuse.

Another approach is OO technique:
```
struct myCustomWriter {
   TagWriter base;
    extraLongLifeCycleData data;
   ...
};
```
This will be a good place to store TMSourceFile.

However, providing the way to return "the result of parsing a file" from ctags can make a client code a bit simple.

Reply:
I was thinking about passing custom data into createTags() rather than the data being returned by it - returning data is just a subset of what we may want to do and when we allow users to pass an arbitrary pointer inside, they have the flexibility to do about anything with the passed data. This is a pretty common pattern in various libraries, e.g.

https://developer.gnome.org/glib/stable/glib-Arrays.html#g-array-sort-with-data

where the user data is passed to the compare_func. In a similar way, createTags() could look like createTags(..., void *user_data) and it would store the user data somewhere, e.g. the extraLongLifeCycleData which would then be accessible from the writer.

Reply:
> createTags(..., void *user_data)

Oh, I see.  I will update our changes here. Thank you.